### PR TITLE
i#2210 clean call: Implement parameter preparation on ARM and AArch64.

### LINKED ISSUE #2210: AArchXX clean calls handle far too few use cases, blocking tool development
The AArchXX implementation of insert_parameter_preparation() is missing some key functionality:

```
/* FIXME i#1551, i#1569: we only implement naive parameter preparation,
 * where args are all regs or immeds and do not conflict with param regs.
 */
```

This is a big enough missing port feature to call for a separate issue.  This is blocking tool development on AArchXX, even simple samples: https://github.com/DynamoRIO/dynamorio/pull/2174

Xref original issues #1551, #1569

### PR DESCRIPTION
**Pull request by egrimley**
The implementation should efficiently handle any (reasonable) number
of arguments that are immediate integers or full-size registers.
Test cases are added in cleancallparams.dll.c.

Fixes #2210

### CODE REVIEW

File: core/arch/aarchxx/mangle.c [hunk 1]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+#endif
+    }
+
+    /* Initialise regs[], which encodes the contents of parameter registers. */
+    for (i = 0; i < num_regs; i++) {
+        if (opnd_is_immed_int(args[i]))
+            regs[i] = -1;
+        else {
+            reg_id_t reg = opnd_get_reg(args[i]);
+            regs[i] = -2;
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: See above too: this -1, -2 too stuff needs comments explaining, or, better, to use an enum type or something

File: core/arch/aarchxx/mangle.c [hunk 2]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+                         opnd_is_reg(*arg0) ? *arg0 : opnd_create_reg(DR_REG_XZR)));
+                }
+            } else if (i + 1 < n && opnd_is_reglike(*arg1)) {
+                /* str x(...), [sp, #((i + 1) * XSP_SZ)] */
+                PRE(ilist, instr, instr_create_1dst_1src
+                    (dcontext, OP_str,
+                     opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                           (i + 1) * XSP_SZ, OPSZ_PTR),
+                     opnd_is_reg(*arg1) ? *arg1 : opnd_create_reg(DR_REG_XZR)));
+            }
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Where are the immeds and xsp?

Update: they are down below: but I don't understand why?  IMHO it would be much clearer to insert them here rather than splitting up where these stack params are written.

File: core/arch/aarchxx/mangle.c [hunk 3]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+                                           opnd_create_reg(DR_REG_XSP),
+                                           OPND_CREATE_INT32(stack_inc)));
+        for (i = 0; i < n; i++) {
+            opnd_t arg = args[NUM_REGPARM + i];
+            if (opnd_is_reglike(arg)) {
+                /* str r(...), [sp, #(i * 4)] */
+                PRE(ilist, instr, XINST_CREATE_store
+                    (dcontext, opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                                     i * 4, OPSZ_PTR), arg));
+            }
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Ditto: confusing to split up where the stack params are stored: why have the reg params in between?

File: core/arch/aarchxx/mangle.c [hunk 4]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+                if (regs[i] != -2)
+                    --usecount[regs[i]];
             }
-        } else {
-            /* FIXME i#1551, i#1569: we only implement naive parameter preparation,
-             * where args are all regs or immeds and do not conflict with param regs.
-             */
-            ASSERT_NOT_IMPLEMENTED(false);
+            regs[i] = i;
+            break;
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: I don't see the reason for the break for any reason except the else above where usecount is decremented: why start over for the prior two cases?

File: core/arch/aarchxx/mangle.c [hunk 5]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+        }
+    } while (i < num_regs);
+
+    /* From now on it is safe to use LR as a temporary. */
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Maybe more explanation would make this clearer (and asserts?): I assume the logic is that the only thing to worry about is LR as a param, and since it's not a calling conv arg reg, it would have be moved into one by the loop above.

File: core/arch/aarchxx/mangle.c [hunk 6]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+            if (regs[i] != i)
+                break;
+        }
+        if (i >= num_regs)
+            break;
+        first = i;
+        PRE(ilist, instr, XINST_CREATE_move(dcontext, opnd_create_reg(DR_REG_LR),
+                                            opnd_create_reg(regparms[i])));
+        do {
+            tmp = regs[i];
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: OK I think I understand what it's doing, shifting one into LR and everything else one down, then repeat, but comments on the approach would help, and on whether there might be a more efficient method but it would be much more complex and not worthwhile, etc.

File: core/arch/aarchxx/mangle.c [hunk 7]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+            PRE(ilist, instr,
+                XINST_CREATE_move(dcontext, opnd_create_reg(regparms[i]),
+                                  tmp == first ? opnd_create_reg(DR_REG_LR) :
+                                  opnd_create_reg(regparms[tmp])));
+            regs[i] = i;
+            i = tmp;
+        } while (tmp != first);
+    }
+
+    /* Set up stack arguments that are (non-zero) constants or SP. */
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: See above: IMHO these should be up by the other stack args

File: core/arch/aarchxx/mangle.c [hunk 8]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
-        DODEBUG({
-            uint j;
-            /* check no reg used by arg conflicts with regparms */
-            for (j = 0; j < i; j++)
-                ASSERT_NOT_IMPLEMENTED(!opnd_uses_reg(args[j], regparms[i]));
-        });
     }
-    return 0;
+
+    return (uint)stack_inc;
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Any reason it can't be declared as a uint?

File: core/arch/mangle_shared.c [hunk 9]
```diff
@@ -464,14 +464,17 @@ insert_meta_call_vargs(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
...
             INSTR_CREATE_lea(dcontext, opnd_create_reg(REG_XSP),
                              opnd_create_base_disp(REG_XSP, REG_NULL, 0,
                                                    stack_for_params, OPSZ_lea)));
-#elif defined(ARM)
-        /* FIXME i#1551: NYI on ARM */
-        ASSERT_NOT_IMPLEMENTED(false);
+#elif defined(AARCHXX)
+        PRE(ilist, instr, INSTR_CREATE_add(dcontext, opnd_create_reg(REG_XSP),
+                                           opnd_create_reg(REG_XSP),
+                                           OPND_CREATE_INT32(stack_for_params)));
```
Comment (derekbruening) [inline core/arch/mangle_shared.c]: We can remove all the ifdefs and use XINST_CREATE_add(), right?

File: core/arch/aarchxx/mangle.c [hunk 10]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+    uint num_regs = num_args < NUM_REGPARM ? num_args : NUM_REGPARM;
+    signed char regs[NUM_REGPARM];
+    int usecount[NUM_REGPARM];
+    ptr_int_t stack_inc = 0;
+    uint i, j;
 
-    ASSERT(num_args == 0 || args != NULL);
-    /* FIXME i#1551, i#1569: we only support limited number of args for now. */
-    ASSERT_NOT_IMPLEMENTED(num_args <= NUM_REGPARM);
+    /* We expect every arg to be an integer or a full-size register. */
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Forgot to mention this in the review summary as this is a more major thing: current docs do not say memory args are disallowed.  Is that what we want to do, update the docs to disallow them?

File: core/arch/aarchxx/mangle.c [hunk 11]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
 }
 #endif /* !AARCH64 */
 
+/* Return true if opnd is a register, but not XSP, or immediate zero on AArch64. */
+static bool
+opnd_is_reglike(opnd_t opnd)
+{
+    return ((opnd_is_reg(opnd) && opnd_get_reg(opnd) != DR_REG_XSP) ||
+            IF_X64_ELSE((opnd_is_immed_int(opnd) && opnd_get_immed_int(opnd) == 0),
+                        false));
```
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: There's no "||false" in there...

File: core/arch/aarchxx/mangle.c (hunk 10 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: I don't think it would be hard to add memory operands, but then they should also be a test of memory operands, which would probably mean porting "cleancall", so probably that should be a separate commit.

File: core/arch/aarchxx/mangle.c [hunk 12]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
-                                   ilist, instr_get_next(mark), NULL, NULL);
-        } else if (opnd_is_reg(args[i])) {
-            opnd_t arg = opnd_create_reg(reg_to_pointer_sized(opnd_get_reg(args[i])));
-            if (opnd_get_reg(arg) == DR_REG_XSP) {
-                instr_t *loc = instr_get_next(mark);
-                PRE(ilist, loc, instr_create_save_to_tls
-                    (dcontext, regparms[i], TLS_REG0_SLOT));
-                insert_get_mcontext_base(dcontext, ilist, loc, regparms[i]);
-                PRE(ilist, loc, instr_create_restore_from_dc_via_reg
+        ASSERT(opnd_is_immed_int((args[i])) ||
```
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: So for this commit I should probably do a CLIENT_ASSERT that also accepts memory args, followed by an ASSERT_NOT_IMPLEMENTED that rejects the memory args.

File: core/arch/aarchxx/mangle.c [hunk 13]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+    }
+
+    /* Set up stack arguments that are registers (not SP) or zero (on AArch64). */
+    if (num_args > NUM_REGPARM) {
+        uint n = num_args - NUM_REGPARM;
+        stack_inc = ALIGN_FORWARD(n, 2) * XSP_SZ;
+#ifdef AARCH64
+        for (i = 0; i < n; i += 2) {
+            opnd_t *arg0 = &args[NUM_REGPARM + i];
+            opnd_t *arg1 = &args[NUM_REGPARM + i + 1];
```
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: I thought you were referring to the loop, at first: that's safe because n = num_args - NUM_REGPARM. But &args[NUM_REGPARM + i + 1] should also be fine because you're allowed to take the address of the element just beyond the end of the array.

File: core/arch/aarchxx/mangle.c (hunk 2 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: There's no guarantee of there being a free register at this point. See the comment: /* From now on it is safe to use LR as a temporary. */

File: core/arch/aarchxx/mangle.c [hunk 14]
```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
...
+        PRE(ilist, instr, INSTR_CREATE_sub(dcontext, opnd_create_reg(DR_REG_XSP),
+                                           opnd_create_reg(DR_REG_XSP),
+                                           OPND_CREATE_INT32(stack_inc)));
+        for (i = 0; i < n; i++) {
+            opnd_t arg = args[NUM_REGPARM + i];
+            if (opnd_is_reglike(arg)) {
+                /* str r(...), [sp, #(i * 4)] */
+                PRE(ilist, instr, XINST_CREATE_store
+                    (dcontext, opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0,
+                                                     i * 4, OPSZ_PTR), arg));
```
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: That's weird. I thought I'd searched for all suspicious mentions of "4" and "8"...

File: core/arch/aarchxx/mangle.c (hunk 3 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: We do the reg params in order to make sure that LR is available as a scratch register.

File: core/arch/aarchxx/mangle.c (hunk 6 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: A rotation of n registers is implemented with (n + 1) moves in the obvious way. I don't think there's a better way.

File: core/arch/aarchxx/mangle.c (hunk 7 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: No guarantee of a free register earlier.

File: core/arch/aarchxx/mangle.c (hunk 8 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: It might involve a lot of casts to keep Microsoft compilers happy. Though Microsoft compilers won't be looking at this code in the short term...

File: core/arch/mangle_shared.c (hunk 9 above)
Comment (egrimley) [inline core/arch/mangle_shared.c]: Yes, probably. I don't understand why the X86 code is done how it is.

File: core/arch/aarchxx/mangle.c (hunk 11 above)
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: ?  For non-x64 it is "||\n    false"

File: core/arch/mangle_shared.c (hunk 9 above)
Comment (derekbruening) [inline core/arch/mangle_shared.c]: You mean why it's OP_lea?  Because OP_add clobbers the flags.

Comment (egrimley) [inline core/arch/mangle_shared.c]: But XINST_CREATE_add() will generate an OP_lea on X86?

Comment (derekbruening) [inline core/arch/mangle_shared.c]: Please look at the docs for XINST_CREATE_add versus XINST_CREATE_add_s and the implementations.  We deliberately set it up that way after some discussion.

File: core/arch/aarchxx/mangle.c (hunk 10 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: I'm slightly unsure about which memory operands should be accepted:

constant offset from a register: Yes.
register offset from a register: A pain to implement but hard to justify excluding.
REL_ADDR_kind: Is this useful? (It would be AArch32-only.)
MEM_INSTR_kind: Does that make any sense here?
ABS_ADDR_kind: Does not exist for ARM/AArch64.

File: core/arch/aarchxx/mangle.c (hunk 4 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: The obvious alternative, using a bool, is more LOC, but perhaps you prefer it.

File: core/arch/aarchxx/mangle.c (hunk 8 above)
Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: I like it to be a signed type locally because I pass -stack_inc to functions that take an int. Perhaps the return type should be be ptr_int_t or ptr_uint_t: it's an increment to a pointer type and in theory one could have 2^32 arguments on a 64-bit system.

File: core/arch/aarchxx/mangle.c (hunk 7 above)
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: A comment should explain this: it was not obvious.

Comment (egrimley) [inline core/arch/aarchxx/mangle.c]: I've added a comment near the start of the function: /* The strategy here is to first set up ... (Yes, I do split infinitives!)

File: core/arch/aarchxx/mangle.c [hunk 15]
```diff
@@ -596,20 +595,34 @@ insert_parameter_preparation(dcontext_t *dcontext, instrlist_t *ilist, instr_t *
     ptr_int_t stack_inc = 0;
     uint i, j;
 
-    /* We expect every arg to be an integer or a full-size register. */
+    /* We expect every arg to be an immediate integer, a full-size register,
+     * or a simple memory reference (NYI).
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Nice to have i# and XXX or FIXME

File: core/arch/aarchxx/mangle.c [hunk 16]
```diff
@@ -596,20 +595,34 @@ insert_parameter_preparation(dcontext_t *dcontext, instrlist_t *ilist, instr_t *
...
     for (i = 0; i < num_args; i++) {
-        ASSERT(opnd_is_immed_int((args[i])) ||
-               (opnd_is_reg(args[i]) && reg_get_size(opnd_get_reg(args[i])) == OPSZ_PTR));
-    }
+        CLIENT_ASSERT(opnd_is_immed_int((args[i])) ||
+                      (opnd_is_reg(args[i]) &&
+                       reg_get_size(opnd_get_reg(args[i])) == OPSZ_PTR) ||
+                      opnd_is_base_disp(args[i]),
+                      "insert_parameter_preparation: bad argument type");
+        ASSERT_NOT_IMPLEMENTED(!opnd_is_base_disp(args[i]));
```
Comment (derekbruening) [inline core/arch/aarchxx/mangle.c]: Best to have XXX or FIXME with i#
